ducktape/tests/result_store.py

```diff
@@ -0,0 +1,175 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from ducktape.utils.util import ducktape_version
+
+import os
+import json
+import sqlite3
+import time
+
+
+def create_test_datum(result):
+
+    return {
+        "timestamp": time.time(),
+        "ducktape_version": ducktape_version(),
+        "session_id": result.session_context.session_id,
+        "description": result.description,
+        "test_id": result.test_context.test_id,
+        "module_name": result.test_context.module_name,
+        "cls_name": result.test_context.cls_name,
+        "function_name": result.test_context.function_name,
+        "injected_args": result.test_context.injected_args,
+        "run_time_sec": result.run_time,
+        "services": result.test_context.services,
+        "status": "pass" if result.success else "fail",
+        "error_msg": result.summary,
+        "data": result.data
+    }
+
+
+def create_session_datum(session_datum):
+    return session_datum
+
+
+class TestKey(object):
+    """
+    A TestKey identifies a particular test with a particular set of arguments
+    """
+
+    @staticmethod
+    def from_test_context(test_context):
+        return TestKey(test_context.module_name,
+                                test_context.cls_name,
+                                test_context.function_name,
+                                test_context.injected_args)
+
+
+    @staticmethod
+    def from_datum(datum):
+        return TestKey(datum["module_name"],
+                                datum["cls_name"],
+                                datum["function_name"],
+                                datum["injected_args"])
+
+    def __init__(self, module_name, cls_name, function_name, injected_args={}):
+        self.module_name = module_name
+        self.cls_name = cls_name
+        self.function_name = function_name
+        self.injected_args = injected_args
+
+    def __repr__(self):
+        return "~~".join(
+            [self.module_name,
+             self.cls_name,
+             self.function_name,
+             json.dumps(self.injected_args, separators=(',', ':'))])
+
+
+class FileSystemResultStore(object):
+    DB_NAME = "test_data.db"
+
+    def __init__(self, root_results_dir):
+        """Assume key is a string representing a path relative to root result directory"""
+        self.root_results_dir = root_results_dir
+        self.db_file = os.path.join(self.root_results_dir, "test_data.db")
+        self._bootstrap_store()
+
+    def _bootstrap_store(self):
+        with sqlite3.connect(self.db_file) as connection:
```

Comment:
Wow, a real DB! I though for sure this was going to be a bunch of pickled objects.